### PR TITLE
fix: local data retention not reclaiming disk space & not auto-refreshing after cleanup

### PR DESCRIPTION
**Pull request by Anshgrover23**
### Description
Fixed multiple bugs in the local data retention feature that caused disk usage to stay unchanged after cleanup and the UI to appear frozen with no feedback.      

- Tests passing: 
Screenshot 2026-04-14 182213

- Working: 
Screenshot 2026-04-14 182324

https://github.com/user-attachments/assets/fca577ec-cc0a-4987-a79d-319a811e746c

### Bugs fixed

  - PRAGMA incremental_vacuum was a silent no-op — SQLite requires auto_vacuum=INCREMENTAL to be set at DB creation for this pragma to work; it was never set, so   
  deleted data never actually freed disk space. Replaced with a proper VACUUM via a new db.vacuum() method on the write pool.
  - last_cleanup only updated after all batches finished — for users with months of data this takes minutes, making the UI look completely frozen. Now marks cleanup
   as started immediately so the UI shows activity right away.
  - "Clean up now" polled status once after 3s then stopped — changed to poll every 4s for up to 2 minutes so progress is visible as batches complete.
  - Disk usage card never refreshed automatically — RetentionSettings and DiskUsageSection were fully isolated; added onCleanupComplete callback prop so the disk   
  usage section auto-recalculates in the same polling loop as retention status.

### Files changed

 - crates/screenpipe-db/src/db.rs — Added pub async fn vacuum() using the write pool
  - crates/screenpipe-engine/src/retention.rs — Replaced PRAGMA incremental_vacuum → db.vacuum(); mark last_cleanup at cleanup start
  - apps/screenpipe-app-tauri/components/settings/retention-settings.tsx — Continuous 4s polling for 2min after "clean up now"; added onCleanupComplete prop        
  - apps/screenpipe-app-tauri/components/settings/disk-usage-section.tsx — Pass refetch as onCleanupComplete to RetentionSettings
  - crates/screenpipe-db/tests/retention_test.rs — New file, 10 integration tests covering delete, orphan cleanup, vacuum, and boundary conditions

### CODE REVIEW

**Comment by Anshgrover23** [comment]
@louis030195 Can I get a review on this one ?

**Comment by louis030195** [comment]
@Anshgrover23 before the vacuum - can you somehow make the tauri app cehck if the db is larger than 6 gb if so send ntoficaiton saying thatits doing osme cleanup or something

since its going to block write to db for a bit

for a 20 gb db it takes 3-4 minutes (tested locally)

**Comment by nehaaprasad** [comment]
### AI Code Trust — ship readiness

**Score:** 97 · **Verdict:** SAFE · **Model:** deterministic+groq-v1

No blocking issues detected by automated checks. There are unresolved issues with interval cleanup, VACUUM approach, and empty database checks

**Top issues:**
- **logic** (medium) — at apps/screenpipe-app-tauri/components/settings/retention-settings.tsx:185, an interval is created but not properly cleaned up if an error occurs
  _at_ [`apps/screenpipe-app-tauri/components/settings/retention-settings.tsx:185`](https://github.com/screenpipe/screenpipe/blob/d41904d3d1f9f30597ab7c37ff2fdd2494d76f99/apps/screenpipe-app-tauri/components/settings/retention-settings.tsx#L185)
- **logic** (medium) — at crates/screenpipe-db/tests/retention_test.rs:280, the test_get_oldest_timestamp_empty_db function does not check if the database is empty before calling get_oldest_timestamp
  _at_ [`crates/screenpipe-db/tests/retention_test.rs:280`](https://github.com/screenpipe/screenpipe/blob/d41904d3d1f9f30597ab7c37ff2fdd2494d76f99/crates/screenpipe-db/tests/retention_test.rs#L280)
- **logic** (low) — at crates/screenpipe-engine/src/retention.rs:402, the full VACUUM approach is used but no error handling is provided if auto_vacuum is not set to INCREMENTAL
  _at_ [`crates/screenpipe-engine/src/retention.rs:402`](https://github.com/screenpipe/screenpipe/blob/d41904d3d1f9f30597ab7c37ff2fdd2494d76f99/crates/screenpipe-engine/src/retention.rs#L402)

File: apps/screenpipe-app-tauri/components/settings/retention-settings.tsx [hunk 1]
```diff
@@ -175,8 +179,15 @@ export function RetentionSettings() {
         throw new Error(err.error || "failed to trigger cleanup");
       }
       toast({ title: "cleanup triggered" });
-      // poll status after a short delay
-      setTimeout(fetchStatus, 3000);
+      // Poll retention status + disk usage every 4s for up to 2 minutes
+      // so both sections update automatically as batches complete
+      let polls = 0;
+      const pollId = setInterval(async () => {
```
Comment (nehaaprasad) [inline apps/screenpipe-app-tauri/components/settings/retention-settings.tsx:185]: **logic** · medium

at apps/screenpipe-app-tauri/components/settings/retention-settings.tsx:185, an interval is created but not properly cleaned up if an error occurs

File: crates/screenpipe-db/tests/retention_test.rs [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    #[tokio::test]
+    async fn test_get_oldest_timestamp_empty_db() {
+        let db = setup_db().await;
+        let result = db
+            .get_oldest_timestamp()
+            .await
+            .expect("get_oldest_timestamp failed");
+        assert!(result.is_none(), "expected None for empty DB");
+    }
+
```
Comment (nehaaprasad) [inline crates/screenpipe-db/tests/retention_test.rs:280]: **logic** · medium

at crates/screenpipe-db/tests/retention_test.rs:280, the test_get_oldest_timestamp_empty_db function does not check if the database is empty before calling get_oldest_timestamp

File: crates/screenpipe-engine/src/retention.rs [hunk 3]
```diff
@@ -387,10 +396,12 @@ async fn do_local_cleanup(db: &Arc<DatabaseManager>, cutoff: DateTime<Utc>) -> a
...
             warn!("retention: orphan chunk cleanup failed: {}", e);
         }
-        // Reclaim disk space — without VACUUM, SQLite keeps deleted pages
-        info!("retention: running incremental vacuum to reclaim disk space");
-        if let Err(e) = db.execute_raw_sql("PRAGMA incremental_vacuum(1000)").await {
-            warn!("retention: incremental vacuum failed: {}", e);
+        // Reclaim disk space — VACUUM rewrites the DB file and shrinks it.
+        // PRAGMA incremental_vacuum requires auto_vacuum=INCREMENTAL (not set),
+        // so it would be a silent no-op. Full VACUUM is the correct approach.
+        info!("retention: running vacuum to reclaim disk space");
```
Comment (nehaaprasad) [inline crates/screenpipe-engine/src/retention.rs:402]: **logic** · low

at crates/screenpipe-engine/src/retention.rs:402, the full VACUUM approach is used but no error handling is provided if auto_vacuum is not set to INCREMENTAL
